feat: expose underlying *spanner.Client

egonelbre opened the pull request:
Sometimes the exposed features of go-sql-spanner are not sufficient to get the necessary response. This adds an escape hatch to access the underlying client.

Fixes #312

olavloite commented on driver.go:
```diff
@@ -479,6 +479,10 @@ type SpannerConn interface {
 	// that committed successfully. The timestamp is in the local timezone.
 	CommitTimestamp() (commitTimestamp time.Time, err error)
+
+	// UnderlyingClient returns the underlying client to the database.
+	// This should not be used together with transactions and batching.
```
```suggestion
	// UnderlyingClient returns the underlying Spanner client for the database.
	// The client cannot be used to access the current transaction or batch on
	// this connection. Executing a transaction or batch using the client that is
	// returned, does not affect this connection.
	// Note that multiple connections share the same Spanner client. Calling
	// this function on different connections to the same database, can
	// return the same Spanner client.
```

olavloite commented on driver.go:
```diff
@@ -542,6 +546,16 @@ const (
 )
 
+func (c *conn) UnderlyingClient() (*spanner.Client, error) {
+	if c.inTransaction() {
+		return nil, spanner.ToSpannerError(status.Error(codes.FailedPrecondition, "cannot access *spanner.Client when in a transaction"))
+	}
+	if c.batch != nil {
+		return nil, spanner.ToSpannerError(status.Error(codes.FailedPrecondition, "cannot access *spanner.Client with an active batch"))
+	}
```
Purely from a technical point of view, I don't think we need this restriction, although I understand it from a conceptual viewpoint; the client that is returned cannot be used to interact with the current transaction on the connection. There is however nothing wrong with using the client that is returned to execute other transactions or queries.

egonelbre replied:
Oh sure, I would be completely fine removing the limitation.

egonelbre replied:
Any opinion whether it should be:

```
UnderlyingClient() (client *spanner.Client, err error)

or

UnderlyingClient() (client any, err error)
```

olavloite replied:
I would say `UnderlyingClient() (client *spanner.Client, err error)`. Anyone calling this method does so with the intention of using Spanner-specific features, so declaring that the function returns a Spanner-specific type seems reasonable to me.

olavloite reviewed (commented):
Thanks for adding this. I would lean towards removing the restriction that the client cannot be gotten when there is an active transaction/batch, but I would like to hear whether you had any specific reasons for adding the restriction?

egonelbre commented:
> Thanks for adding this. I would lean towards removing the restriction that the client cannot be gotten when there is an active transaction/batch, but I would like to hear whether you had any specific reasons for adding the restriction?

@olavloite I removed the restriction. I initially thought the batching was actively using the client somehow. I think I addressed all the pending questions.
